src/Symfony/Component/Security/Core/Security.php

```diff
@@ -69,4 +77,35 @@ public function getToken(): ?TokenInterface
     {
         return $this->container->get('security.token_storage')->getToken();
     }
+
+    public function autoLogin(UserInterface $user, AuthenticatorInterface $authenticator = null): void
+    {
+        $request = $this->container->get('request_stack')->getCurrentRequest();
+
+        if (null === $authenticator) {
+            $firewall = $this->container->get('security.firewall.map')->getFirewallConfig($request);
+
+            if (null === $firewall) {
+                throw new LogicException('No firewall found as the current route is not covered by any firewall.');
+            }
+            $firewallName = $firewall->getName();
+
+            if (!\array_key_exists($firewallName, $this->authenticators) || 0 === \count($this->authenticators[$firewallName])) {
+                throw new LogicException(sprintf('No authenticators found for the firewall "%s".', $firewallName));
+            }
+            $firewallAuthenticators = $this->authenticators[$firewallName];
```

Comment:
once my previous suggestion applied, this will be a `ServiceProviderInterface` instance. You can use `getProvidedServices()` to access the authenticator identifiers for the selection logic below